core/src/main/scala/com.mediative.amadou/Stage.scala

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2017 Mediative
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mediative.amadou
+
+import org.apache.spark.sql._
+import scala.util.{ Failure, Try }
+
+sealed trait Stage[-I, +T] { self =>
+  def name: String
+
+  def map[U](f: T => U): Stage[I, U] = new Stage[I, U] {
+    override def name = self.name
+    override def run(ctx: Stage.Context[I]): Stage.Result[U] = self.run(ctx).map(f)
+  }
+
+  def flatMap[U](f: T => Stage[T, U]): Stage[I, U] = new Stage[I, U] {
+    override def name = self.name
+    override def run(ctx: Stage.Context[I]): Stage.Result[U] =
+      self.run(ctx).flatMap(data => f(data).run(ctx.withValue(data)))
+  }
+
+  def andThen[U](s: Stage[T, U]): Stage[I, U] = flatMap(_ => s)
+  def ~>[U](s: Stage[T, U]): Stage[I, U] = flatMap(_ => s)
```

Comment:
I know these are grouped together visually but it would still be great to add a little Scaladoc here to say that `~>` is exactly the same thing as `andThen`.